src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -256,18 +239,20 @@ private void CheckInvariants()
                 newFrozenSourceGeneratedDocumentState);
         }
 
-        private SolutionState CreatePrimarySolution(
+        public SolutionState WithNewWorkspace(
             string? workspaceKind,
             int workspaceVersion,
-            HostWorkspaceServices services)
+            SolutionServices services)
         {
             if (workspaceKind == WorkspaceKind &&
-                workspaceVersion == _workspaceVersion &&
+                workspaceVersion == WorkspaceVersion &&
                 services == Services)
             {
                 return this;
             }
 
+            // Note: this will potentially have problems if the workspace services are different, as some services
+            // get locked-in by document states and project states when first constructed.
```

Comment:
I see the comment moved but just to confirm: we aren't making this any worse than we already are?

Reply:
No, nothing changed. It's the same insanity as before :)

Reply:
"Same insanity" is a quality bar I can get behind!